src/Confluent.Kafka/Error.cs

```diff
@@ -51,6 +83,8 @@ public Error(ErrorCode code)
             Code = code;
             reason = null;
             IsFatal = code == ErrorCode.Local_Fatal;
+            IsRetriable = !IsFatal;
```

Comment:
Huhm, I'm not sure this is completely correct and I think it would be better to be consistent with the underlying librdkafka errors in that we do not make all existing errors retriable.

Ignoring txnAbortable since it is specific to transactions, there are three error states:
 * fatal - is it not meaningful to continue operating the current client instance.
 * retriable - the error indicated a temporary issue, retry the operation
 * neither - the error is most likely permanent, but that doesn't mean the application needs to treat is a fatal error.

Most non-retriable errors are not fatal, the fatality in these cases is up to the program logic.
E.g., is an UnknownTopic for a QueryWatermarkOffsets() call fatal for the instance? No. Is it fatal for the application? Depends on the program logic.

To be consistent with the underlying librdkafka I think it is better to leave these as false.

The way I'm thinking with APIs that return the new rd_kafka_error_t type is instead of documenting specific error codes, the API docs should simply state what the application should based on the three error states above.
Not all APIs might return a retriable error, and skipping it from the doc string saves people from wrapping all calls in retry-loops.

Reply:
> The way I'm thinking with APIs that return the new rd_kafka_error_t type is instead of documenting specific error codes, the API docs should simply state what the application should/could do based on the three error states above.

yes, but they should be `true` i think (as i have them, the negation of fatal), not `false`, to indicate to the application that yes, this is retriable, and yes, this is txn abortable. UnknownTopic for QueryWatermarkOffsets or QueueFull on Produce - both should have IsRetriable and IsTxnAbortable `true` (and fatal `false`).

I was definitely thinking this was a dangerous approximation to be making, and also considered making these constructors obsolete so the dangerous approximation is not exposed outside the insides of the library (where it can be fixed as needed). that was more work though so I didn't. I might do that now.

Reply:
txnAbortable should definitely not be set by default, it does not even have meaning outside the transactional API and setting it to false will just be confusing.

As for retriable: I'd argue that UnknownTopic is not a retriable error, if the topic was not automatically created by referencing it the first time it is unlikely that another call to the same API will get better results, and it just tricks users into repeatedly calling APIs, which could have side effects.

QueueFull, Timed out, Transport failure, on the other hand are retriable, but we should not blindly base the retriable flag on the error code itself, but a combination of error and API (and underlying protocol request), and that is something only librdkafka should do since  the bindings don't have enough information to make that decision.

So, to play it safe I'd urge us to ease into this transition by only setting these flags where we're completely sure they are connect, which is currently on the transactional API.

Reply:
And, it is better to error on the side of caution.

Reply:
> txnAbortable should definitely not be set by default, it does not even have meaning outside the transactional API and setting it to false will just be confusing.

This makes complete sense to me - if you *are* in a transaction, you're ok to abort it (if you're not, just ignore, it's clear what to do). The opposite of what you say is the most natural to me. The difference in the way we reationalize this underscores that this is a crap API. What happens if you do QueryWatermarkOffsets inside a transaction? IsTxnAbortable should be `true`, it won't be. Many other examples. People will take these values literally. Marking QueueFull as not-retriable will confuse people.

For now, I've just hidden these properties, which keeps the Error class API the same as before and not confusing. we can work out what to do later. This is a step up from providing non-meaningful values.

Reply:
txn_abortable() doesn't mean it is okay but optional to abort(), but that the transaction needs to be aborted for the producer to recover, and this will only be signalled from transactional APIs.
It doesn't make sense for other APIs (QueryWmarks..) to set that flag since it has nothing to do with transactions.

I think a TransactionError sub-class would be a good solution to avoid confusion here though and not keep txnAbortable on the main Error class.

> People will take these values literally. Marking QueueFull as not-retriable will confuse people.

Yes, but we need to do this organically to not change semantics later, and the easiest way to do that is to document that retriable(), et.al, are only relevant for the transactional API at this point.
Example:
https://github.com/confluentinc/confluent-kafka-go/blob/master/kafka/producer.go#L855
> // Returns nil on success or an error object on failure.
> // Check whether the returned error object permits retrying
> // by calling `err.(kafka.Error).IsRetriable()`, or whether an abortable
> // or fatal error has been raised by calling
> // `err.(kafka.Error).IsTxnAbortable()` or `err.(kafka.Error).IsFatal()`
> // respectively.

I.e., it explicitly tells the user how to handle the returned error object.

Reply:
> txn_abortable() doesn't mean it is okay but optional to abort(), but that the transaction needs to be aborted for the producer to recover, and this will only be signalled from transactional APIs.

The name implies to me the semantics I assumed. If this were called `txn_requires_abort()`, I would have guessed the semantics correctly. change the name?

> `TransactionError`

maybe - for sure I'm thinking there would be a `KafkaTransactionException`. I think i would just put the `TxnRequiresAbort` property on that rather than create another error class as well (but would implement and see how it feels).

can you clarify exactly the intended semantics of `is_retriable` - i believe your plan that this be applicable everywhere?

If so, I think that should be a property on `Error`. What I can do is make this throw a `NotImplementedException` by default if accessed. For the cases where we have an `error_t`, i can populate with the value from that. For selected other cases I can make it behave correctly (maybe just do the QueueFull on produce as it's common). Alternatively, just leave this off.

> and the easiest way to do that is to document that retriable(), et.al, are only relevant for the transactional API at this point.

people won't read the docs - they will see the property exists in intellisense and make assumptions about how it works in every context. the more we require people read the docs the more difficult and time consuming it is to use and the more mistakes people will make.